File: server-scala/src/main/scala/com/anontown/ports/ClientLoader.scala
package com.anontown.ports
import zio.IO
import com.anontown.AtError
import com.anontown.entities.client.{Client, ClientId}

trait ClientLoader {
  def load(id: ClientId): IO[AtError, Client];
  def loadMany(ids: List[ClientId]): IO[AtError, List[Client]];
}

trait ClientLoaderComponent {
  val clientLoader: ClientLoader;
}

File: server/packages/server/src/adapters/client-loader/client-loader.ts
import * as DataLoader from "dataloader";
import { Client } from "../../entities/index";
import { IAuthContainer, IClientLoader, IClientRepo } from "../../ports/index";
import { loader } from "../loader-helper";

export class ClientLoader implements IClientLoader {
  loader: DataLoader<string, Client>;

  constructor(clientRepo: IClientRepo, auth: IAuthContainer) {
    this.loader = loader(ids =>
      clientRepo.find(auth.getTokenMasterOrNull(), { id: ids }),
    );
  }

  load(id: string) {
    return this.loader.load(id);
  }

  loadMany(ids: Array<string>) {
    return this.loader.loadMany(ids);
  }
}
